sdk/resource/resource.go

```diff
@@ -17,30 +17,53 @@
 package resource
 
 import (
-	"reflect"
+	"sort"
+	"strings"
 
 	"go.opentelemetry.io/otel/api/core"
 )
 
 // Resource describes an entity about which identifying information and metadata is exposed.
 type Resource struct {
-	labels map[core.Key]core.Value
+	str    string
+	sorted []core.KeyValue
+	keys   map[core.Key]struct{}
 }
 
 // New creates a resource from a set of attributes.
 // If there are duplicates keys then the first value of the key is preserved.
 func New(kvs ...core.KeyValue) *Resource {
-	res := &Resource{
-		labels: map[core.Key]core.Value{},
-	}
+	res := &Resource{keys: make(map[core.Key]struct{})}
 	for _, kv := range kvs {
-		if _, ok := res.labels[kv.Key]; !ok {
-			res.labels[kv.Key] = kv.Value
+		// First key wins.
+		if _, ok := res.keys[kv.Key]; !ok {
+			res.keys[kv.Key] = struct{}{}
+			res.sorted = append(res.sorted, kv)
 		}
 	}
+	sort.Slice(res.sorted, func(i, j int) bool {
+		return res.sorted[i].Key < res.sorted[j].Key
+	})
+	res.str = buildResourceString(res.sorted)
```

Comment:
If the purpose here is to cache a potentially expensive string construction I would suggest deferring it until `String()` is invoked for the first time.

```suggestion
```